### PR TITLE
[RLlib] Add eval worker sub-env fault tolerance test.

### PR DESCRIPTION
**Pull request by sven1977**
- Fix the `restart_failed_sub_environments` feature for multi-agent settings (incl. multi-agent w/ `remote_worker_env=True`).
- Fix the missing support for `restart_failed_sub_environments` in case a sub-environment fails during its `reset()` call (this was previously not supported).
- Adds logging messages around sub-environments getting restarted: "trying to restart ..." and "restarted successfully".
- Change the existing PG sub-environment fault tolerance learning test to one that also adds eval workers (parallel eval and training).
- Add more PG learning tests for setting `restart_failed_sub_environments` for multi-agent cartpole (crashing once in a while) and multi-agent cartpole with remote worker envs (crashing once in a while).
- Adds parallel eval and training option to custom_eval.py example script and add to BUILD.

TODO: Documentation PR that better explains what happens under the hood in case `restart_failed_sub_environments=True` and an environment fails during step or reset.

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [ ] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [ ] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: rllib/evaluation/sampler.py [hunk 1]
```diff
@@ -1174,6 +1174,7 @@ def _do_policy_eval(
             # have already been changed (mapping fn stay constant
             # within one episode).
             episode = active_episodes[eval_data[0].env_id]
+            assert not episode.is_faulty
```
[the remark this answers is not in the document]
Comment (sven1977) [reply rllib/evaluation/sampler.py:1177]: We do, we test this fault-tolerance behavior (restarting only the failed sub-environment) on a special version of CartPole that crashes with some probability every once in a while during `step()`.
There is a learning(!) test case for PG with these settings, including eval workers in `tuned_examples/pg/cartpole-crashing-pg.yaml`

File: rllib/tuned_examples/pg/cartpole-crashing-pg.yaml [hunk 2]
```diff
@@ -2,8 +2,8 @@ cartpole-crashing-pg:
     env: ray.rllib.examples.env.cartpole_crashing.CartPoleCrashing
     run: PG
     stop:
-        episode_reward_mean: 150.0
-        timesteps_total: 120000
+        evaluation/episode_reward_mean: 180.0
```
Comment (sven1977) [inline rllib/tuned_examples/pg/cartpole-crashing-pg.yaml:5]: Make sure, learning success is measurable by eval workers.

File: rllib/utils/debug/memory.py [hunk 3]
```diff
@@ -83,13 +83,16 @@ def check_memory_leaks(
         action_sample = action_space.sample()
 
         def code():
+            horizon = algorithm.config["horizon"] or float("inf")
```
Comment (sven1977) [inline rllib/utils/debug/memory.py:86]: Respect horizon setting in this utility.

File: rllib/examples/custom_eval.py [hunk 4]
```diff
@@ -76,6 +76,7 @@
 from ray.rllib.utils.test_utils import check_learning_achieved
 
 parser = argparse.ArgumentParser()
+parser.add_argument("--evaluation-parallel-to-training", action="store_true")
```
Comment (sven1977) [inline rllib/examples/custom_eval.py:79]: Add parallel option to this script and add setting this option to tests in BUILD.

File: rllib/algorithms/tests/test_algorithm.py [hunk 5]
```diff
@@ -347,6 +350,30 @@ def test_no_env_but_eval_workers_do_have_env(self):
         bc.train()
         bc.stop()
 
+    def test_eval_workers_on_infinite_episodes(self):
```
Comment (sven1977) [inline rllib/algorithms/tests/test_algorithm.py:353]: New test to check for proper behavior if eval workers are set to run by episodes, but episodes dont terminate.

File: rllib/algorithms/algorithm_config.py [hunk 6]
```diff
@@ -175,6 +175,7 @@ def __init__(self, algo_class=None):
         self.evaluation_interval = None
         self.evaluation_duration = 10
         self.evaluation_duration_unit = "episodes"
+        self.evaluation_sample_timeout_s = 180.0
```
Comment (sven1977) [inline rllib/algorithms/algorithm_config.py:178]: New config value. Triggers a meaningful warning if breached with various things to try to fix the timeouts.

File: rllib/algorithms/algorithm.py [hunk 7]
```diff
@@ -2339,6 +2358,14 @@ def _run_one_evaluation(
                     "recreate_failed_workers"
                 ),
             )
+
+        # Add number of healthy evaluation workers after this iteration.
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:2362]: This metric was missing.

File: rllib/algorithms/algorithm.py [hunk 8]
```diff
@@ -753,7 +753,8 @@ def duration_fn(num_units_done):
                     env_steps_this_iter += batch.env_steps()
                 metrics = collect_metrics(
                     self.workers.local_worker(),
-                    keep_custom_metrics=self.config["keep_per_episode_custom_metrics"],
+                    keep_custom_metrics=eval_cfg["keep_per_episode_custom_metrics"],
+                    timeout_seconds=eval_cfg["metrics_episode_collection_timeout_s"],
```
Comment (sven1977) [inline rllib/algorithms/algorithm.py:757]: bug: timeout setting was not passed to eval collect-metrics call.

**Review by kouroshHakha — APPROVED** [review]
Thanks @sven1977 LGTM.

**Pull request merged**